networking/src/Ntp/Client.hs

```diff
@@ -77,280 +80,209 @@ data NtpClient = NtpClient
-spawnNtpClient settings ncStatus =
-    withSocketsDoLifted $
+spawnNtpClient settings ncStatus = do
+    logInfo "starting"
     bracket (mkSockets settings) closeSockets $ \sock -> do
         cli <- mkNtpClient settings ncStatus sock
 
-        addrs <- catMaybes <$> mapM (resolveHost $ socketsToBoolDescr sock)
-                                    (ntpServers settings)
+        addrs <- catMaybes <$> traverse resolveNtpHost (ntpServers settings)
         when (null addrs) $ throwM NoHostResolved
-        () <$ startReceive cli `concurrently`
-              startSend cli addrs `concurrently`
-              logInfo "Launched NTP client"
-  where
+        startReceive cli
+            `concurrently_` sendLoop cli addrs
+            `concurrently_` logInfo "started"
+    where
+    closeSockets :: Sockets -> IO ()
     closeSockets sockets = do
-        logInfo "NTP client is stopped"
-        forM_ (socketsToList sockets) close
-    resolveHost sockDescr host = do
-        maddr <- resolveNtpHost host sockDescr
-        case maddr of
-            Nothing   -> do
-                logWarning $ sformat ("Host "%shown%" is not resolved") host
-                pure Nothing
-            Just addr -> do
-                logInfo $ sformat ("Host "%shown%" is resolved: "%shown) host addr
-                pure $ Just addr
+        foldEitherOrBoth $ bimap fn fn sockets
+        logInfo "stopped"
 
+    fn :: Last (WithAddrFamily t Socket.Socket) -> IO ()
+    fn (Last sock) = Socket.close $ runWithAddrFamily sock
 
 -- | Run Ntp client in a seprate thread, return a mutable cell which holds
 -- `NtpStatus`.
 withNtpClient :: MonadIO m => NtpClientSettings -> m (TVar NtpStatus)
 withNtpClient ntpSettings = do
-    ntpStatus <- newTVarIO NtpSyncPending
-    _ <- liftIO $ async (spawnNtpClient ntpSettings ntpStatus)
-    return ntpStatus
-
--- | Run without Ntp client.
-withoutNtpClient :: MonadIO m => (TVar NtpStatus -> m a) -> m a
-withoutNtpClient f = newTVarIO NtpSyncUnavailable >>= f
-
--- | Start client, wait for a while so that most likely it ticks once
--- and stop it.
-ntpSingleShot
-    :: NtpClientSettings
-    -> TVar NtpStatus
-    -> IO ()
-ntpSingleShot ntpSettings ncStatus =
-    () <$ timeout (ntpResponseTimeout ntpSettings) (spawnNtpClient ntpSettings ncStatus)
-
--- Store created sockets.
--- If system supports IPv6 and IPv4 we create socket for IPv4 and IPv6.
--- Otherwise only one.
-data Sockets
-    = IPv4Sock !Socket
-    | IPv6Sock !Socket
-    | BothSock !Socket !Socket
-    deriving Show
-
-socketsToList :: Sockets -> [Socket]
-socketsToList (BothSock s1 s2) = [s1, s2]
-socketsToList (IPv4Sock s1)    = [s1]
-socketsToList (IPv6Sock s1)    = [s1]
+    liftIO $ logInfo "withNtpClient"
+    ncStatus <- newTVarIO NtpSyncPending
+    -- using async so the ntp thread will be left running even if the parent
+    -- thread finished.
+    _ <- liftIO $ async (spawnNtpClient ntpSettings ncStatus)
```

Comment:
How many times is this function likely to be called? Does the `spawnNtpClient` ever terminate?

If it never terminates,  how do we make sure we don't get hundreds, thousands or even more?

If if just runs, does it job and then terminates, it would be nice to have a comment to that effect.

Reply:
It just runs once, a comment should be in place indeed.